jupyterhub/spawner.py

```diff
@@ -522,6 +543,17 @@ def get_args(self):
         args.extend(self.args)
         return args
 
+    def run_pre_spawn_hook(self):
+        """
+        run a pre spawn hook function which you can define in your jupyter_config.py
+        :return: void
+        """
+        if (callable(self.pre_spawn_hook)):
+            try:
+                self.pre_spawn_hook(self)
+            except Exception as e:
+                self.log.error("pre_spawn_hook failed with exception: %s", str(e))
```

Comment:
If you use `self.log.exception`, it will log an error message and the traceback, so you wouldn't need `str(e)` here.